core/templates/dev/head/domain/topic/TopicUpdateService.js

```diff
@@ -230,6 +228,61 @@ oppia.factory('TopicUpdateService', [
           // Get the current change list.
           var currentChangeList = UndoRedoService.getChangeList();
           var indicesToDelete = [];
+          var oldSubtopicIdSkillIdMap = {};
+          var index = null;
+          var skillIds = subtopic.getSkillIds();
+          for (var i = 0; i < currentChangeList.length; i++) {
+            var changeDict =
+              currentChangeList[i].getBackendChangeObject();
+            if (oldSubtopicIdSkillIdMap[changeDict.skill_id] === undefined) {
+              if (changeDict.new_subtopic_id === subtopicId) {
+                // Record the origin subtopic id for the skill.
+                oldSubtopicIdSkillIdMap[changeDict.skill_id] =
+                  changeDict.old_subtopic_id;
+                index = i;
+              }
+            } else {
+              if (changeDict.cmd === CMD_REMOVE_SKILL_ID_FROM_SUBTOPIC) {
+                // If the skill was removed from all subtopics at some point,
+                // modify that change dict, so that it was as if it was removed
+                // from its recorded origin.
+                changeDict.subtopic_id =
+                  oldSubtopicIdSkillIdMap[changeDict.skill_id];
+              } else {
+                // At every move operation, set old subtopic id as the recorded
+                // old subtopic id.
+                changeDict.old_subtopic_id =
+                  oldSubtopicIdSkillIdMap[changeDict.skill_id];
+              }
+              currentChangeList[i].setBackendChangeObject(changeDict);
+              // Remove the previous move operation that was related to the
+              // skill id since, the old subtopic id of the current change dict
+              // is set to the actual initial subtopic id.
+              currentChangeList.splice(index, 1);
```

Comment:
It's a bit icky to modify the thing you're iterating over. I suggest building up a replacementChangeList instead and substituting it in at the end (as long as it's OK to break references to dicts within that list).

Reply:
Done, appended these indices also to indicesToDelete.